trojsten/diplomas/sources.py

```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
+
+import re
+import json
+from abc import abstractmethod
+from bs4 import BeautifulSoup
+
+try:
+    from urllib2 import urlopen
+except ImportError:
+    from urllib.request import urlopen
+
+from django.template import loader
+from django.utils import timezone
+
+
+class AbstractSource(object):
+    name = 'abstract_source'
+
+    def __init__(self):
+        pass
+
+    def handle_request(self, request, **kwargs):
+        """
+        Rendered component on the page is capable of making a request to the backend
+        (for cases when obtaining participant data is computationally expensive or they
+        are queried from the server). This method provides such interface.
+
+        :param request:
+        :param kwargs:
+        :return:
+        """
+        return []
+
+    @abstractmethod
+    def render(self, **kwargs):
+        """
+        Method outputs HTML representation of the source plugin.
+
+        :param kwargs:
+        :return: Rendered HTML that will be displayed as a component on page
+        """
+        return ''
+
+
+class FileUpload(AbstractSource):
+    name = 'file_upload'
+
+    def render(self, **kwargs):
+        template = loader.get_template('trojsten/diplomas/sources/fileupload.html')
+        return template.render()
+
+
+class Naboj(AbstractSource):
+    """
+    Base class for Naboj result webpage scraper.
+    """
+
+    name = 'naboj'
+
+    @staticmethod
+    def compose_url(pattern, *args, **kwargs):
+        url_args = []
+        for k, v in kwargs.items():
```

Comment:
Netreba znovu vynaliezat koleso:
https://docs.python.org/3.3/library/urllib.parse.html#urllib.parse.urlencode

Pozor, python 2 ma urlib.urlencode

Da sa to pouzit takto: urlencode(dict) a vyrobi to url encoded string k=v oddeleny & plus to odescapuje blbe znaky.